### PR TITLE
Premature convergence

### PR DESCRIPTION
**Pull request by pire-TO**
I think we can merge

### CODE REVIEW

File: cifopy/charles/charles.py [hunk 1]
```diff
@@ -108,9 +124,34 @@ def evolve(self, gens, select, crossover, mutate, co_p, mu_p, elitism):
...
+                if gen%50==0:
+                    print(f'Best Individual: {min(self, key=attrgetter("fitness"))} Gen: {gen}')
+    
+    def reshuffle(self):
+        n = len(self)
+        new_pop = []
+        if self.optim == "max":
+            new_pop.append(deepcopy(max(self.individuals, key=attrgetter("fitness"))))
+        elif self.optim == "min":
+            new_pop.append(deepcopy(min(self.individuals, key=attrgetter("fitness"))))
```
Comment (mariushes) [inline cifopy/charles/charles.py:140]: This is to keep the best individual, right? If elitism is used, I think we do it twice. Do you think it's a problem?

File: cifopy/charles/charles.py [hunk 2]
```diff
@@ -108,9 +124,34 @@ def evolve(self, gens, select, crossover, mutate, co_p, mu_p, elitism):
...
+        
+        
+        keep = sample(range(n), k=n//2)
+        #for how it's implemented the best could appear twice. Don't think it's a problem but we can avoid it
+        #half of the times you keep the best twice, could also be useful. write in report
+        for i in keep:
+            new_pop.append(deepcopy(self.individuals[i])) #to test, not sure
```
Comment (mariushes) [inline cifopy/charles/charles.py:147]: Do we actually need to deep copy them? The individuals won't be used anymore anyway right?

File: cifopy/charles/selection.py [hunk 3]
```diff
@@ -8,14 +9,26 @@ def fps(population):
...
+        fitness = [i.fitness for i in population]
+        """
+        if prem:
+            #Do we want to save this variance? Maybe we can plot how it changes through generation
+            var = np.var(fitness)
+            if var/population.initial_var<0.05:
+                print("Premature convergence! Reshuffling the deck")
+                population.reshuffle()
+                #Re-evaluate fitness since population changed
+                fitness = [i.fitness for i in population]"""
```
Comment (mariushes) [inline cifopy/charles/selection.py]: Can we delete this?

File: cifopy/charles/charles.py (hunk 1 above)
Comment (pire-TO) [reply cifopy/charles/charles.py:140]: Elitism comes to a different stage, if we didn't do this we would potentially lose the best. A different issue in the immediate following lines where we save half of the current population. In this way we could have the best individual twice (50% of the times, there should be a comment in the code). I don't think this is a big issue but we maybe is best to change it in order to be more consistent.

File: cifopy/charles/charles.py (hunk 2 above)
Comment (pire-TO) [reply cifopy/charles/charles.py:147]: I did not want to make any mess with pointers and individuals, I think also Dave deepcopies often. Probably would work the same, but anyway few lines later self.individuals is overwritten by new_pop, so the old individuals will be discarded anyway.

File: cifopy/charles/selection.py (hunk 3 above)
Comment (pire-TO) [reply cifopy/charles/selection.py]: Yes for sure, it was the previous implementation. Forgot to delete it :)

**Pull request merged**